Themes: Switch data fetching to Redux

ockham opened the pull request:
_Notes by @mcsf:_

### Context

What started as an experimental branch for a number of Redux-related changes across Themes coalesced into a substantial refactor of our data fetching components, which are now Redux-`connect`ed. This means that this PR moves Themes entirely to Redux.

### Decisions

- `ThemesListFetcher`, `ActivatingTheme`, and `CurrentTheme` (i.e. the data component) are kept as generic data components that relay application state to their children.
- Selectors are given much greater importance, as they allow us to simplify the data model by avoiding forms of data duplication — notably, avoiding storage of derived data next to essential data.
- Use of selectors allowed to altogether eliminate one action: `SEARCH_THEMES`, which was basically a _virtual_ action (read: _hack_) to trigger client-side filtering of themes, a feature specific to Jetpack sites.

### Next

- Stores, and helper functions in reducers are now obsolete, and are removed in #1363 to keep this PR's size down.
- We need to move files to different locations in order to comply with #751. This PR is big enough already, so that will be done separately.
- As we progress, more derived data should be able to be eliminated from the global state's `themes` branch.
- Some features of the Showcase invoke non-Themes action creators (namely, purchases and customization). This PR uses the `dispatch()` function from `context.reduxStore` in the relevant sections to dispatch the corresponding Redux actions. Another possible approach would have been to have a layer of Redux middleware that dispatches certain actions to the singleton Flux architecture.
- Memoization of selectors is something very useful to have, and trivial to implement thanks to immutability. However, there are *server-side-rendering*-specific concerns; specifically:
  * To avoid user-specific data leakage between Node requests, most memoized selectors would have to be bound to a particular request and be GC'd afterwards. This would work the same way that our global Redux store on the server is always bound to a request.
  * For performance improvements, it could be very advantageous that some non-user-specific selectors were capable of cross-request memoization, though care should be given to ensure said selectors's memory usage doesn't grow unintentionally.

### Bugs

- [x] Jetpack sites: searching refetches the themes list when it doesn't need to, though this only happens for the first search.
- [x] Jetpack sites: searching blurs the search field.

### Testing

QA should focus around navigation of the Showcase, switching between sites, including both WP.com and Jetpack sites, searching, per-tier filtering, triggering `InfiniteList`'s fetching, and switching between Showcase and other sections.

ockham commented:
Just tried to stick this into #846, to provide data to the themes list we render there, server-side. This could work okayish, but ideally, I think we should use the Redux-based `themes-list-fetcher` on the client side too, in order not to duplicate code. So maybe that could be the goal of this PR?

ockham commented:
@ehg @seear Join the party :-)

mcsf commented:
> I think we should use the Redux-based themes-list-fetcher on the client side too, in order not to duplicate code. So maybe that could be the goal of this PR?

Yeah, that's where I'm thinking of heading!

mcsf commented:
#338 opened a nice precedent for client-side usage of a Redux global store, so… seems like the pieces are falling together, we can work with that momentum.

ehg commented:
I noticed that search/tier selection was funky, 0aca329 is a possible fix :P

mcsf commented:
I meant to fix it, thanks for taking care of it :).

ehg commented on client/boot/index.js:
```diff
@@ -87,7 +88,15 @@ function setUpContext( layout ) {
+		//		console.log( 'Flux -> Redux' );
+		//		reduxStore.dispatch( action );
+		//	} );
+		//}() );
```
Interesting. Did this work? :)

I'm thinking we probably don't need to integrate the two dispatchers now, as the Redux transition is getting some momentum behind it. We should check we're not relying on any 'external' actions, although from what I can remember, we're not.

mcsf replied:
> Interesting. Did this work? :)

It did. :)

> we probably don't need to integrate the two dispatchers now

I agree. I pushed it for the sake of communicating the exploration process, i.e. as a way of saying that this in particular had been tried.

> We should check we're not relying on any 'external' actions, although from what I can remember, we're not.

Yeah, IIRC only the reverse happens, and that's something we can handle at the Redux middleware level.

ehg commented on shared/lib/create-redux-store/index.js:
```diff
@@ -9,8 +10,11 @@ import { createStore, applyMiddleware } from 'redux';
 	return applyMiddleware(
-		thunkMiddleware
+		thunkMiddleware,
+		loggerMiddleware
```
Could be useful in a separate PR, enabled via an `env` var to `make run`?

mcsf replied:
Yep. This got in for WIP debugging purposes (see the bit about Flux→Redux re-dispatching).

ockham replied:
We also need to add our `analyticsMiddleware` here, don't we?

mcsf replied:
Added — you might want to test it yourself as a double-check that it works.

ehg commented on shared/components/data/themes-list-fetcher/index.jsx:
```diff
@@ -49,76 +49,103 @@ let ThemesListFetcher = React.createClass( {
-		if ( page > 1 && ! loading && lastPage ) {
-			this.onLastPage && this.onLastPage();
-		}
+		fetchNextPage( site );
```
I got kind of confused between the AC `fetchNextPage` and the fetcher's `fetchNextPage` function. Perhaps namespace/rename?

ehg commented on shared/components/data/themes-list-fetcher/index.jsx:
```diff
@@ -49,76 +49,103 @@ let ThemesListFetcher = React.createClass( {
+			incrementThemesPage,
+			fetchThemes,
+			fetchJetpackThemes,
+		} = this.props;
```
Is this going to get stale `props`?

mcsf replied:
Hm, do you have a scenario in mind? These will be passed by `connect`'s wrapping component, which should be smart enough.

ehg commented on shared/my-sites/themes/controller.js:
```diff
@@ -39,18 +40,20 @@ var controller = {
-					analyticsPageTitle,
-					'Themes'
-				)
+			React.createElement( ReduxProvider, { store: context.reduxStore }, () => {
```
Possibly add a comment like `// TODO: Remove when using React 0.14`

ehg commented:
This is looking great. I removed some dead code, and rebased.

<sub>The Jetpack search is quite amusing atm ;)</sub>

mcsf commented:
Thanks for your feedback and fixes. I just pushed substantial changes that should make this PR a lot simpler and ship-worthy. Subsequent feedback and some testing would be greatly appreciated. :)

> The Jetpack search is quite amusing atm ;)

Heh, indeed. Hopefully, it's looking better now.

mcsf commented:
On derived data and selectors: the changes brought around `searchJetpackThemes`/`getFilteredThemes` are a good illustration of what I've said in the past about Redux allowing us to significantly trim the `themes` branch of the global state.

Our data needs for `themes` are likely too simple to justify the 4 subreducers we have for it. Improving that isn't the goal of this PR, but it's something to be aware of.

ehg commented:
> the changes brought around searchJetpackThemes/getFilteredThemes are a good illustration of what I've said in the past about Redux allowing us to significantly trim the themes branch of the global state

Wow, yes. This looks much nicer. Perhaps update the PR description with what you guys have done, and how it's better? I figure this would be a good example to give to others, on how to `reduxify`.

mcsf commented:
> Perhaps update the PR description with what you guys have done, and how it's better? I figure this would be a good example to give to others, on how to reduxify.

Indeed. I've rewritten the PR description and added the NR label.

ockham commented:
Quick note from testing -- purchasing a theme in single-site mode doesn't seem to update the `CurrentTheme` component.

ockham commented:
How well do our `flux-actions` (that use my `combineStores()` helper at an intermediate step -- and hence Flux stores created from reducers via `lib/store`'s `createReducerStore()`) work with the global Redux store (created from Redux's own `createStore()`)? Do we have to remove them, and use our Redux actions directly everywhere?

ockham commented on shared/my-sites/themes/controller.js:
```diff
@@ -39,18 +40,21 @@ var controller = {
-				)
+			// TODO: Remove function when using React 0.14
+			React.createElement( ReduxProvider, { store: context.reduxStore }, () => {
+				return React.createElement( ThemesComponent, {
```
Can't we shorten `() => { return x; }` to `() => x` ?

mcsf replied:
Absolutely.

mcsf commented:
> How well do our flux-actions (that use my combineStores() helper at an intermediate step -- and hence Flux stores created from reducers via lib/store's createReducerStore()) work with the global Redux store (created from Redux's own createStore())? Do we have to remove them, and use our Redux actions directly everywhere?

Yeah, that's the unsolved bit. Short of porting all of Themes to Redux at once, I'm not sure what the least bad approach is for now. A quick hack would be to take anything coming in through the Redux dispatcher and re-dispatch into the Flux dispatcher; that way, data gets duplicated into the Themes Flux stores, and other bits can keep consuming that. Ideas are very welcome.

ockham commented:
> Short of porting all of Themes to Redux at once

`ag` gives me 10 occurrences of `flux-actions` in the code. Rolling up my sleeves.

ockham commented:
> Quick note from testing -- purchasing a theme in single-site mode doesn't seem to update the `CurrentTheme` component.

Should be fixed now.

ockham commented on shared/components/data/themes-list-fetcher/index.jsx:
```diff
@@ -1,44 +1,45 @@
-
-function getThemesState() {
+import Constants from 'lib/themes/constants';
+import * as allActions from 'lib/themes/actions';
```
Hmm, do we really want `allActions`? We could just `import { query, fetchNextPage } from 'lib/themes/actions'` and `connect()` to `bindActionCreators.bind( null, { query, fetchNextPage } )`, couldn't we?

mcsf replied:
Indeed, that sounds better. Don't know what led me to the current approach.

mcsf replied:
I think I remember now, it was likely because of this:

<img width="603" alt="screen shot 2015-12-08 at 02 14 56" src="https://cloud.githubusercontent.com/assets/150562/11646040/9d6853bc-9d51-11e5-9acb-f2385ef2e4cb.png">

Between adding an `eslint-disable-line no-shadow` comment and doing `{ query: boundQuery, fetchNextPage: boundFetchNextPage } = this.props`, I don't know which is uglier. But reverting to calling `this.props.query` seems like a shame. In a way, the `allActions / pick` approach _kind_ of made sense. Thoughts? :)

mcsf replied:
Bit the bullet in e84c571. :]

ockham commented:
TODO (probably for a later PR, in order to not unnecessarily bloat this one any more):

* [ ] Rewrite tests in `shared/lib/themes` to directly test reducers instead of stores
* [ ] Kill `shared/lib/themes/stores`
* [ ] Remove obsolete helper functions from reducers, now that they are in `selectors.js`
* [ ] Profit.

mcsf commented on client/components/data/current-theme/index.js:
```diff
@@ -1,65 +1,67 @@
+import { fetchCurrentTheme } from 'lib/themes/actions';
+import { getCurrentTheme } from 'lib/themes/selectors';
+
+function getCurrentThemeState( state, { site } ) {
```
Totally up to you, but I notice that other data components just go with a uniform `getState` name, since it's clear, in the context of each file, what the function should do.

ockham replied:
Fixed.

mcsf commented on client/components/data/current-theme/index.js:
```diff
@@ -1,65 +1,67 @@
-		this.setState( {
-			currentTheme: CurrentThemeStore.getCurrentTheme( this.props.site.ID )
-		} );
+	refresh( props ) {
```
I suggest `nextProps`

mcsf replied:
Actually, scratch that. :)

ockham commented:
Pushed a commit to demonstrate what removing stores and modifying tests needs to look like. Feel free to ignore -- we can always branch :-)

mcsf commented:
> `ag` gives me 10 occurrences of flux-actions in the code. Rolling up my sleeves.

> [purchasing a theme in single-site mode doesn't seem to update the CurrentTheme component] […] Should be fixed now.

Nice work! :) Everything looks great. I'm only tempted to tweak things in `shared/my-sites/themes/main.jsx` to make things a bit easier on the eyes, but maybe I should look at that in the morning, if at all.

> Pushed a commit to demonstrate what removing stores and modifying tests needs to look like. Feel free to ignore -- we can always branch :-)

Again, looks great — clean testing — and I can't wait to kill the stores, but it feels like a separate PR, no?

ehg commented on shared/my-sites/themes/main.jsx:
```diff
@@ -67,7 +71,7 @@ var Themes = React.createClass( {
 		const site = this.props.sites.getSelectedSite();
 		if ( site.jetpack ) {
-			Action.customize( theme, site );
+			this.props.dispatch( Action.customize( theme, site ) );
```
I wonder if it makes sense having the global Redux store as a global variable, so we don't have to pass these props around, if we're not using `connect()`

ockham replied:
Yeah, probably. Seems to get pretty ugly otherwise.

ockham replied:
OTOH, adding `reduxStore` to `window` isn't going to play nice with SSR :-/

mcsf replied:
> OTOH, adding reduxStore to window isn't going to play nice with SSR :-/

That was my first reaction. Passing `dispatch` here seems good enough™ to me, for a first iteration at least.

ockham commented:
> Pushed a commit to demonstrate what removing stores and modifying tests needs to look like. Feel free to ignore -- we can always branch :-)

Rebased, and while at it, removed that commit again. Flux stores removal is now in #1363.

ockham commented:
Noting that the `isJetpack()` call inside the `themes-list` reducer is still causing a bit of trouble.

mcsf commented:
> Noting that the isJetpack() call inside the themes-list reducer is still causing a bit of trouble.

Oh? But `isJetpack` should be available as an action param — _cf_ [receive action](https://github.com/Automattic/wp-calypso/pull/840/files#diff-22104692e1a08563a6558230f3d10614R104) — so no need to use (and retrofit) selectors inside the reducer, right?

ockham commented:
> Oh? But `isJetpack` should be available as an action param — _cf_ [receive action](https://github.com/Automattic/wp-calypso/pull/840/files#diff-22104692e1a08563a6558230f3d10614R104) — so no need to use (and retrofit) selectors inside the reducer, right?

I'm afraid one of the main uses of the `last-query` reducer is to set `isJetpack` during `INCREMENT_THEMES_PAGE` -- which is the value we want to check during `themes-list`'s `RECEIVE_THEMES` handling, as opposed to the action param available at that point :-/

ockham commented:
Hrmph, the only solution I can seem to think of right now is moving `isJetpack` to the `themes-list` reducer. Any alternatives?

ockham commented:
Removed the 'In Progress' label, since I think this is ready for review now.

ockham commented:
For completeness' sake, noting that I fixed the `isJetpack` issue by 

> having a thunk on the action side, calling `getState`, selecting the bool via the `isJetpack` selector?

as @mcsf suggested in an internal discussion.

ockham commented:
Oh, and I think the Bugs listed in the PR description should be fixed now, too -- @mcsf can you confirm?

ockham commented:
Some notes from a walkthrough with @mcsf @ehg and @seear today:

* It seems a bit arbitrary what goes in `selectors`. Some of those helpers we could move to the components where they are used. As for the actual selectors (that just return a value from a state subtree), we might split them into individual modules (corresponding to the reducers).
* We might want to rename `getState()` in our data components to something less confusing, e.g. `getPropsFromState()`.
* Those props that are passed to data components via `connect()` aren't really part of the public interface, so we might want to remove the corresponding `propTypes`.
* In the long run, we should get rid of `toJS()`/`toObject()` in our selectors in order to benefit from  comparing Immutable.js objects directly.
* I'll try to `connect()` `themes-selection` and `main` in order to get rid of those messy stores and dispatches being passed down the component hierarchy.
* In `customize` and `upgrades`, we should add a note on those Redux dispatches for later.

(Most of these issues we're likely to address at a later iteration in order to not bloat this PR any further.)

ockham commented:
> Those props that are passed to data components via `connect()` aren't really part of the public interface, so we might want to remove the corresponding `propTypes`.

FWIW, `propTypes` are listed for that type of props at http://rackt.org/redux/docs/basics/UsageWithReact.html ¯\\\_(ツ)_/¯

ockham commented:
In order to get rid of some more of the ugliness that is `main.jsx`, we'll need to refactor `theme-options`. In `main.jsx`, we're now passing props `selectedAction` (a string), `actions`, and `getOptions` to `ThemesSiteSelectorModal`. Ideally, I think it should be just a `selectedAction` function -- something that is currently computed inside `ThemesSiteSelectorModal`. Urgh.

ockham commented:
> We should add a note on Redux dispatches for later in `customize`

Or migrate `customize/actions` to Redux (thunks invoking the Flux `Dispatcher`, that is) 'cause passing `dispatch` down the component hierarchy like I'm doing now is oh. so. ugly.

mcsf commented on client/components/data/activating-theme/index.js:
```diff
@@ -1,49 +1,49 @@
+		children: React.PropTypes.element.isRequired,
+		// The following props aren't part of this module's public interface,
+		// since they are provided by the `connect()` call that is wrapped around
+		// `ActivatingThemeData` before it is exported.
```
This doesn't feel like the way to go; I mean, as we progress, so many components will be `connect`ed — will we paste a three-line comment in each one's `propTypes`?

My first question would be: is this information really necessary? If it is, there's a subsequent question: can we convey that information through code? For instance, some types could be attached to the plain component, and some others to the connected component. If we prefer not to use code, can the comment be reduced to something like `// @connect`, `// @redux`, etc.?

ockham replied:
> My first question would be: is this information really necessary?

I don't mind really either way -- @ehg and @seear opined that we should remove those props altogether. Somehow, I felt it was giving a somewhat incomplete picture if we did that, so I chose to add those comments. But yeah, scale that to a 100 connected components :-/

> If it is, there's a subsequent question: can we convey that information through code? For instance, some types could be attached to the plain component, and some others to the connected component

Hmm, how would we achieve the latter? I can seem to think only of ES6 class/propTypes-style, i.e. adding a `propTypes` member after it's connected...

> If we prefer not to use code, can the comment be reduced to something like `// @connect`, `// @redux`, etc.?

Sounds okay to me, as long as it doesn't collide with JSDoc statements.

mcsf replied:
> Hmm, how would we achieve the latter? I can seem to think only of ES6 class/propTypes-style, i.e. adding a `propTypes` member after it's connected...

Yeah, that was my rough idea. Inspection via web tools shows that a connected component automatically has `propTypes: { store }` attached to it. On second consideration, I don't know if we should interfere with that mechanism, plus it may not be useful/visible enough.

I'd lean towards the pseudo-JSDoc statement, but I'll welcome anything more correct than that.

mcsf commented:
> Hmm, how would we achieve the latter? I can seem to think only of ES6 class/propTypes-style, i.e. adding a `propTypes` member after it's connected...

Yeah, that was my rough idea. Inspection via web tools shows that a connected component automatically has `propTypes: { store }` attached to it. On second consideration, I don't know if we should interfere with that mechanism, plus it may not be useful/visible enough.

I'd lean towards the pseudo-JSDoc statement, but I'll welcome anything more correct than that.

ockham commented:
Pushed a commit to change them to `// Connected props`, which I'd deem fairly informal -- no need to look up the exact wording for future devs as it'd be okay to put in something equivalent, such as `// connect()` or whatever.

Doing pseudo-JSDoc -- introducing statements that don't exist, and would break JSDoc if we were ever going to use it -- was on second glance just so much against my OCD >:|

mcsf commented:
> Pushed a commit to change them to `// Connected props`, which I'd deem fairly informal -- no need to look up the exact wording for future devs as it'd be okay to put in something equivalent, such as `// connect()` or whatever.

> Doing pseudo-JSDoc -- introducing statements that don't exist, and would break JSDoc if we were ever going to use it -- was on second glance just so much against my OCD >:|

:+1:

seear commented:
I've tested around preview->customize->activate and the _Tracks_ events. Looking good with no apparent regressions. Noting that we don't have a working _Tracks_ event for jetpack search, with or without this PR.

mcsf commented:
:shipit:

ockham commented on shared/my-sites/themes/themes-selection.jsx:
```diff
@@ -46,12 +47,12 @@ const ThemesSelection = React.createClass( {
 			theme: theme.id,
 			results_rank: resultsRank + 1,
-			results: ThemesListStore.get(),
+			results: themesList,
```
Noting here that `ThemesListStore.get()` should have actually read `ThemesListStore.getThemesList()` -- we had overlooked to change `get()` to `getThemesList()` when we started to use `createReducerStore()` to create our stores from reducers. That latter function however returns objects whose `.get()` method returns the entire store state. Thus, we were erroneously setting `results` to the contents of the entire `ThemesListStore` state instead (including query params and whatnot) instead of just the themes list. This has been fixed by this PR.

aduth commented on client/components/data/activating-theme/index.js:
```diff
@@ -1,49 +1,47 @@
 
-module.exports = ActivatingThemeData;
+export default connect(
+	( state, props ) => Object.assign( {},
```
You don't need to assign into the component's own props as you're doing here. Instead, this could be simplified to:

```jsx
export default connect(
	( state, props ) => ( {
		isActivating: isActivating( state ),
		hasActivated: hasActivated( state ),
		currentTheme: getCurrentTheme( state, props.siteId )
	} )
)( ActivatingThemeData );
```

They connected props will be automatically be merged with any other props passed to the component explicitly.

ockham replied:
Thanks for bringing that up! That goes on my things-to-refactor in our Redux subtree list... I need to file an issue for that TODO... :-)
